akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala

```diff
@@ -274,19 +277,54 @@ private[remote] final class EnvelopeBuffer(val byteBuffer: ByteBuffer) {
 
   private def readLiteral(): String = {
     val length = byteBuffer.getShort
-    val bytes = Array.ofDim[Byte](length)
-    byteBuffer.get(bytes)
-    new String(bytes, UsAscii)
+    if (length == 0)
+      ""
+    else {
+      ensureLiteralCharsLength(length)
+      val chars = literalChars
+      val bytes = literalBytes
+      byteBuffer.get(bytes, 0, length)
+      var i = 0
+      while (i < length) {
+        // UsAscii
+        chars(i) = bytes(i).asInstanceOf[Char]
+        i += 1
+      }
+      String.valueOf(chars, 0, length)
+    }
   }
 
   private def writeLiteral(tagOffset: Int, literal: String): Unit = {
-    if (literal.length > 65535)
+    val length = literal.length
+    if (length > 65535)
       throw new IllegalArgumentException("Literals longer than 65535 cannot be encoded in the envelope")
 
-    val literalBytes = literal.getBytes(UsAscii)
     byteBuffer.putInt(tagOffset, byteBuffer.position())
-    byteBuffer.putShort(literalBytes.length.toShort)
-    byteBuffer.put(literalBytes)
+
+    if (length == 0) {
+      byteBuffer.putShort(0)
+    } else {
+      byteBuffer.putShort(literal.length.toShort)
+      ensureLiteralCharsLength(length)
+      val chars = literalChars
+      val bytes = literalBytes
+      // we could use Unsafe to get the internal char array of String, but we are staying safe so far
```

Comment:
agrona+aeron uses Unsafe left and right, so I don't think one more would hurt (if it does, then we are already in trouble).